feature: Implement tenant soft quotas

jshearer opened the pull request:
**Description:**

This introduces per-tenant quotas for collections, captures, materializations, and derivations. At publication time, we compute the change in usage of each of these resources, and compare that to the tenant quota and current number of resources in use. If everything checks out, we update the database with the new resource-in-use counts.

Fixes #741

Example failure:
<img width="1025" alt="Screen Shot 2022-10-25 at 21 30 33" src="https://user-images.githubusercontent.com/4368270/198058600-887e3c8b-7a71-44ba-8d13-2fad9a822676.png">

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/flow/746)
<!-- Reviewable:end -->

jshearer committed:
feature: Implement tenant soft quotas

We also keep track of tenant quota usage!

Fixes #741

TODO: DB migration

psFried commented on crates/agent/src/publications/specs.rs:
```diff
@@ -350,6 +350,238 @@ pub fn validate_transition(
     errors
 }
 
+pub async fn enforce_resource_quotas(
```
This seems like it needs to account for tasks being disabled

jshearer replied:
Done!

jshearer commented:
**TODO:** Switch to `anyhow::Result<Result<HashMap<String, ResourceUsageChanges>>, Vec<Error>>>`

jshearer commented on crates/agent-sql/src/publications.rs:
```diff
@@ -208,6 +208,89 @@ pub async fn resolve_spec_rows(
+            materializations_used,
+            collections_used
+        from tenants
+        where $1 ILIKE tenant || '%';
```
Done!

jshearer committed:
refactor: Lean more heavily on database queries rather than denormalizing these things clientside

jshearer committed:
fix: Take into account disabled specs in quota usage calculation

jshearer commented on crates/agent/src/publications/specs.rs:
```diff
@@ -350,6 +350,238 @@ pub fn validate_transition(
+    // Is there a more idiomatic way to do this with mapping and iterators and stuff?
+
+    for (catalog_name, changes) in catalog_change_map.iter() {
+        let res = agent_sql::publications::find_tenant_for_catalog_name(catalog_name, txn)
```
I rewrote this part

jgraettinger commented on crates/agent-sql/src/publications.rs:
```diff
@@ -208,6 +208,123 @@ pub async fn resolve_spec_rows(
+    pub derivations_quota: i32,
+    pub materializations_quota: i32,
+    pub collections_quota: i32,
+    // These are Option<i64> because even if we add `coalesce(.., 0) to the query
```
nit: instead of doing this, you can tell sqlx in the query that these are non-null by using `foobar as "foobar!"`.

[here's one example](https://github.com/estuary/flow/blob/master/crates/agent-sql/src/publications.rs#L188-L193)

Then you can get rid of `TenantFromDB`

jshearer replied:
Glorious, I tried searching for this but didn't find it

jshearer replied:
Done

jgraettinger commented on crates/agent-sql/src/publications.rs:
```diff
@@ -208,6 +208,123 @@ pub async fn resolve_spec_rows(
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'materialization') as materializations_used,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'collection') as collections_used
+        from tenants
+        join live_specs on live_specs.catalog_name LIKE tenants.tenant || '%' and (live_specs.spec->'shards'->>'disable')::boolean is not true
```
supernit: use starts_with(live_specs.catalog_name, tenants.tenant) for consistency with how we do this everywhere else. This makes code search easier. _Probably_ there's no performance impact if postgres optimizes to this already.

jgraettinger replied:
No action for now, but FYI: `spec` is stored a postgres `JSON` type, and not `JSONB`. It _must_ be this way because `JSONB` will re-order properties and that would break sops macs. But, in this application is means the query must actually parse all of these documents.

A way we can potentially improve this is by creating an index over the pre-extracted values used in this query. It would need to be confirmed, but postgres is smart about using a covering index to avoid needing to do the parsing work.

jshearer replied:
Done

jgraettinger commented on crates/agent-sql/src/publications.rs:
```diff
@@ -208,6 +208,123 @@ pub async fn resolve_spec_rows(
+        from tenants
+        join live_specs on live_specs.catalog_name LIKE tenants.tenant || '%' and (live_specs.spec->'shards'->>'disable')::boolean is not true
+        where tenants.tenant = ANY($1)
+        group by tenants.tenant, tenants.captures_quota, tenants.materializations_quota, tenants.derivations_quota, tenants.collections_quota;
```
this group by should only be over `tenants.tenant`

jshearer replied:
Am I doing something else wrong in the query then? After refactoring, the query looks like this

```sql
        select
            tenants.tenant as name,
            tenants.tasks_quota,
            tenants.collections_quota,
            count(live_specs.catalog_name) filter (where
                live_specs.spec_type = 'capture' or
                live_specs.spec_type = 'materialization' or
                live_specs.spec_type = 'collection' and live_specs.spec->'derivation' is not null
            ) as "tasks_used: i32!",
            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'collection') as "collections_used: i32!"
        from tenants
        join live_specs on live_specs.catalog_name LIKE tenants.tenant || '%' and (live_specs.spec->'shards'->>'disable')::boolean is not true
        where tenants.tenant = ANY($1)
        group by tenants.tenant;
```

And the error I get is

```
error returned from database: column "tenants.tasks_quota" must appear in the GROUP BY clause or be used in an aggregate function
```

jshearer replied:
The intent is to group all of the `live_specs` by tenants.tenant, and then give me the specific `tasks_quota` and `collections_quota` for each grouping (tenant). 

My thought is that including `tasks_quota` and `collections_quota` in the list of fields to group by is okay because there will only ever be one tenant for a particular live_spec, so there will only ever be one "group".

Is there a better way to express this?

jgraettinger replied:
I would use `sum(case when ... )` instead of `count`, where the current `filter` semantics are a case that either returns `1` or `0`.

jgraettinger commented on crates/agent-sql/src/publications.rs:
```diff
@@ -208,6 +208,123 @@ pub async fn resolve_spec_rows(
+}
+
+#[derive(Default)]
+pub struct ResourceUsageChanges {
```
I don't think we should be baking these usage values into the tenants table.

In the future, we could surface the guts of `find_tenant_quotes` as a view `tenants_ext`, which matches the practice we have with `live_specs_ext`, `draft_specs_ext`, etc.

jshearer replied:
Yup, I had removed all of the logic that does the baking-in of usages, just forgot to remove this struct. Done!

jgraettinger commented on crates/agent-sql/src/publications.rs:
```diff
@@ -208,6 +208,123 @@ pub async fn resolve_spec_rows(
+    }
+}
+
+pub async fn find_tenant_quotas(
```
please add test coverage in `agent-sql/tests/publications.rs`

jshearer replied:
done!

jgraettinger commented on supabase/migrations/16_tenants.sql:
```diff
@@ -4,7 +4,17 @@ create table tenants (
-  tenant      catalog_tenant unique not null
+  tenant                  catalog_tenant unique not null,
+
+  captures_quota          integer        not null DEFAULT 10,
```
supernit: we use lowercase keywords. See comment at the top of the first migration `01_json.sql` for details on style.

jshearer replied:
done!

jgraettinger commented on crates/agent/src/publications/specs.rs:
```diff
@@ -350,6 +350,71 @@ pub fn validate_transition(
+}
+
+pub fn tenant_name_for_catalog_name(catalog_name: String) -> anyhow::Result<String> {
+    catalog_name
```
There's a sneaky bug here, which is that this split will drop the trailing '/' and thus won't match any entries of the `tenants` table, which always have a trailing `/`.

An optional suggestion: let the database do more of the heavy lift here. Pass in the spec IDs as `resolve_expanded_rows` does and use a CTE as part of `find_tenant_quotas` which maps that into unique tenants. I think you'll find that this construction is easier to test, and it also removes the possibility of mismatches like this.

jshearer replied:
Done!

jgraettinger commented on crates/agent/src/publications/specs.rs:
```diff
@@ -350,6 +350,71 @@ pub fn validate_transition(
     errors
 }
 
+pub async fn enforce_resource_quotas(
```
Can you add test coverage for this, akin to how directives are tested? Should be pretty feasible to do since these checks occur so early in the publication process.

jshearer replied:
Done!

jgraettinger reviewed (commented):
Looking good, some feedback below.

jshearer commented on crates/agent-sql/src/publications.rs:
```diff
@@ -208,6 +208,123 @@ pub async fn resolve_spec_rows(
+#[derive(Debug)]
+pub struct Tenant {
+    pub name: String,
+    pub captures_quota: i32,
```
Okay cool, done!

jshearer commented:
Awesome, thanks for the detailed feedback! Just pushed a commit that:

* Changes usage metrics to just "tasks", and "collections". Tasks are: captures, materializations, and any collection that has a derivation. Collections are: all collections, including those with derivations
* Rewrote the quota usage query to:
  * Take `live_spec` IDs as input, and compute the tenants from there using a CTE
  * Hint that values aren't optional with `as "foo!"` syntax
  * Use `starts_with(..,..)` instead of `LIKE '%' || '..'`
  * Only group by tenant name. **NOTE/QUESTION:** I had to wrap the quota fields in `max(...)` because we're grouping by tenant. Is there a better way to do this?
* Added a test for `agent-sql/publications/find_tenant_quotas()`
* Added a test for `agent/publications/enforce_resource_quotas()`

jshearer committed:
refactor: Capture previous usage in order to emit more useful error messages

jshearer committed:
refactor: Add a couple more tests, allow publications to go through when over quota so long as the change is net negative, in order to allow publications that disable tasks to go through.

jgraettinger commented on crates/connector_proxy/src/connector_runner.rs:
```diff
@@ -13,8 +15,19 @@ use proto_flow::capture::PullResponse;
+use tokio::task::JoinHandle;
 use tokio_util::io::{ReaderStream, StreamReader};
 
+async fn flatten_join_handle<T, E: std::convert::From<tokio::task::JoinError>>(
```
I think there's a borked merge with upstream -- I'm seeing a bunch of recent changes now being unintentionally reverted in this PR.

jshearer replied:
Oh gross...

jgraettinger reviewed (commented):
Testing work is looking great. Think there's a bad merge with upstream here though.

It looks like you can trivially use `sum` with a predicate to roll up current usage, e.x.:

```
postgres=# select sum((letter = 'a')::int) from (values ('a'), ('b'), ('a'), ('b')) as foo(letter);
 sum 
-----
   2
(1 row)
```

For the CI failure, you may want to have your test case first delete all existing live specs as a CTE, to ensure it's starting from a clean slate. Ideally the test would path against a user's local DB regardless of what they actually have in there.

jshearer committed:
refactor: address PR reviews, add tests

jshearer committed:
chore: add db migration

jshearer commented:
Hoookay, I think I fixed my mistake. It was even groser than a merge, I had accidentally `git reset --soft origin/master` instead of `git reset --soft HEAD~1`, and so rewrote my commit to have a bunch of unnecessary changes from master in it, when all I wanted to do was remove a file that shouldn't have been committed. Anyway, `git reflog` to the rescue, all should be good now.

jshearer committed:
refactor: clean up db before running test

jshearer committed:
fix: Add explicit ordering so tests pass in CI

jshearer committed:
refactor: Improve agent test infrastructure

* Introduce publication tests, with example happy path
* Add `Id::from_hex` to make tests more legible

jshearer committed:
Merge branch 'refactor/improve_agent_tests' into refactor/741_tenant_soft_quotas

jshearer committed:
refactor: Move snapshot inline to make things more legible

jshearer committed:
refactor: Tenant calculation query now with even more CTEs

jshearer commented:
@jgraettinger I rewrote the tenant usage query to use another CTE to avoid grouping by anything more than tenant name. Thoughts?

```sql
with tenant_names as (
            select tenants.tenant as tenant_name
            from tenants
            join live_specs on starts_with(live_specs.catalog_name, tenants.tenant)
            where live_specs.id = ANY($1::flowid[])
            group by tenants.tenant
        ),
        tenant_usages as (
            select
                tenant_names.tenant_name,
                count(live_specs.catalog_name) filter (
                    where
                        live_specs.spec_type = 'capture' or
                        live_specs.spec_type = 'materialization' or
                        live_specs.spec_type = 'collection' and live_specs.spec->'derivation' is not null
                ) as tasks_used,
                count(live_specs.catalog_name) filter (
                    where live_specs.spec_type = 'collection'
                ) as collections_used
            from tenant_names
            join live_specs on
                starts_with(live_specs.catalog_name, tenant_names.tenant_name) and
                (live_specs.spec->'shards'->>'disable')::boolean is not true
            group by tenant_names.tenant_name
        )
        select
            tenants.tenant as name,
            tenants.tasks_quota as "tasks_quota!",
            tenants.collections_quota as "collections_quota!",
            tenant_usages.tasks_used as "tasks_used!",
            tenant_usages.collections_used as "collections_used!"
        from tenant_usages
        join tenants on tenants.tenant = tenant_usages.tenant_name
        order by tenants.tenant;
```

jgraettinger commented:
> I rewrote the tenant usage query to use another CTE to avoid grouping by anything more than tenant name. Thoughts?

I realized I gave you bad advice - you don't need e.x. `task_quota` in a group-by but you _do_ still need some kind of aggregate function wrapping it. Postgres doesn't know if you, for example, might want to sum up the task_quota for _each_ joined live spec (which, to be clear, you don't here), so you have to tell it. I'll suggest using `min()` for this and avoiding the extra CTE.

I was playing with this query and came up with another construction, that I'm offering up here only because I have it already in-hand, and for you to take or leave as you will. `count ... filter` or `sum()` are both fine IMO. Otherwise I think we're veering into SQL code golf 😁 

```sql
with tenants_filtered as (
    -- Select distinct tenants from live_specs participating in the current publication.
    select distinct substring(catalog_name for position('/' in catalog_name)) as tenant
    from live_specs where id = ANY($1::flowid[])
)
select
    t.tenant,
    min(t.tasks_quota) as "tasks_quota!",
    min(t.collections_quota) as "collections_quota!",
    sum((
        (
            s.spec_type = 'capture' or
            s.spec_type = 'materialization' or
            (s.spec_type = 'collection' and s.spec->'derivation' is not null)
        ) and (s.spec->'shards'->>'disable')::boolean is not true
    )::int) as "tasks_used!",
    sum((s.spec_type = 'collection')::int) as "collections_used!"
    from tenants t
    join tenants_filtered f on t.tenant = f.tenant
    join live_specs s on starts_with(s.catalog_name, t.tenant)
    group by t.tenant;
```

jshearer committed:
Merge remote-tracking branch 'origin/master' into refactor/741_tenant_soft_quotas

jshearer committed:
Merge remote-tracking branch 'origin/master' into refactor/741_tenant_soft_quotas

jshearer committed:
test: clean up the diff a little

jgraettinger commented on crates/agent-sql/src/publications.rs:
```diff
@@ -208,6 +208,62 @@ pub async fn resolve_spec_rows(
+#[derive(Debug, Serialize)]
+pub struct Tenant {
+    pub name: String,
+    pub tasks_quota: i32,
```
nit: is there a reason for the i32 / i64 distinction here? Could they be `isize` ?

jshearer replied:
Ummm... I tried a couple of different ways to get them to be the same and I thought sqlx's inference complained about each of them. It looks like making them all `i32` actually _does_ work now, so I'll go with that. FWIW `isize` does not work, sqlx doesn't know how to infer it correctly and complains that:

```
the trait bound `isize: sqlx::Decode<'_, Postgres>` is not satisfied
the following other types implement trait `sqlx::Decode<'r, DB>`:
  f32
  f64
  i16
  i32
  i64
  i8
```

jgraettinger reviewed (approved):
Awesome, really well done, LGTM % question about integer type

jshearer committed:
refactor: Fix unneccesary `i32`/`i64` distinction

jshearer committed:
fix: incorrect casting to `i32`, needed to cast in SQL as well!

* also fix up some snapshots that I forgot
* stable sort so test runs locally

jshearer committed:
Merge remote-tracking branch 'origin/master' into refactor/741_tenant_soft_quotas

jshearer committed:
test: allow `foo` connector to run in test after merging connector image validation!

jshearer committed:
Merge remote-tracking branch 'origin/master' into refactor/741_tenant_soft_quotas

jshearer committed:
fix: add `#[parallel]` annotations to tests
